src/PROPOSAL/detail/PROPOSAL/Secondaries.cxx

```diff
@@ -225,23 +236,30 @@ std::shared_ptr<ParticleState> Secondaries::GetExitPoint(
     return nullptr; // No exit point found
 }
 
-std::shared_ptr<ParticleState> Secondaries::GetClosestApproachPoint(const Geometry& geometry) const
+std::shared_ptr<ParticleState> Secondaries::GetClosestApproachPoint(
+        const Geometry& geometry) const
 {
-    for (unsigned int i = 0; i < track_.size(); i++) {
-        auto sec_pos = track_.at(i).position;
-        auto sec_dir = track_.at(i).direction;
-        if (geometry.DistanceToClosestApproach(sec_pos, sec_dir) <= 0.) {
-            if(std::abs(geometry.DistanceToClosestApproach(sec_pos, sec_dir))
-                        < PARTICLE_POSITION_RESOLUTION)
+   if (track_.size() == 1)
+       return std::make_unique<ParticleState>(track_.front());
+
+    for (unsigned int i = 0; i < track_.size() - 1; i++) {
+        auto pos_i = track_.at(i).position;
+        auto pos_f = track_.at(i+1).position;
+
+        auto displacement = pos_f - pos_i;
+        auto dist_i_f = displacement.magnitude();
+        displacement.normalize();
+
+        auto distance_to_closest_approach
+            = geometry.DistanceToClosestApproach(pos_i, displacement);
+        if (distance_to_closest_approach == dist_i_f) {
```

Comment:
this if-statement should be as precise as the PARTICLE_POSITION_RESOLUTION, see 3 lines below